## Info

Fall of 2018, [Bioinformatics MSc Program](https://bioinfo-grad.gr), University of Crete

This workshop is taught by Christoforos Nikolaou and Savvas Paragkamian from [CG<sup>2</sup> Lab](https://computational-genomics-uoc.weebly.com).

#### Schedule:

The course contains 3 lectures - computer labs and one last meeting for student presentations.

1. Thursday 29/11 14:00-17:00 : [Lecture 1](workshop_0.html) - PC-room (E110)
2. Tuesday 4/12 15:00-18:00 : [Lecture 2](workshop_1.html) - PC-room (E110)
3. Thursday 6/12 14:00-17:00 : [Lecture 3](workshop_2.html) - PC-room (E110)
4. Tuesday 11/12 14:00-17:00 : Q&A - Multimedia Room (E141)
5. Thursday 13/12 14:00-17:00 : [Presentations](#presentations) - PC-room (E110)

## [Lecture 1](workshop_0.html) : Introduction

* what are networks?
* networks in biology
* types of networks
* graph theory basics
* random networks
* scale-free networks

[HTML](workshop_0.html)

## [Lecture 2](workshop_1.html) : Network descriptives

* workspace and workflow introduction
* graphs as mathematical objects and their representation
* import data from database
* network descriptives
* mesoscale analysis of networks
* importance of nodes - centralities

[HTML](workshop_1.html) -- [Rmarkdown file](workshop_1.Rmd)

## [Lecture 3](workshop_2.html) : Network construction from Models and Data

* models for network creation (random, scale free, small world)
* percolation theory - robustness
* network inference from data
* introduction to dynamic systems on networks
* gene ontology annotation
* neural networks vs networks
* state of the art in network science

[HTML](workshop_2.html) -- [Rmarkdown file](workshop_2.Rmd)

## Presentations {#presentations}

For the final examination of the course all students, individually, will present their assignments. They can choose a specific topic they are intersted in or choose from the list below. Students can present their assignments with slides or using markdown with their code. Duration of presentations will be 20 minutes.

#### Structure

**Chapter 1** 

1. Research question
2. Why it is important?
3. Who did what so far?  Review of literature
4. What remains to be done? Why the biologist is not satisfied?

**Chapter 2**

1. Methodology for addressing the problem
2. Necessary theoretical concepts and results

**Chapter 3**

1. Data (origin, reliability, size)
2. Computations Results 
3. Mathematical meaning of the results 

**Chapter 4**

Biological significance of the results 

**Chapter 5**

Conclusions:

1. Mathematical comparison with available results (what is new)
2. Biological comparison with available results (what is new)
3. General comparison with available results (what is new)
4. Perspectives for future work 


#### Featured topics 

1. Generalize assortativity for every centrality measure and redefine it using information theory
2. Compare network community detection methods (Modularity vs Information vs Link-based) and their applications in biology
3. Centrality lethality rule using Gene Ontology centralities
4. Spreading phenomena - dynamics Stefanos 
5. Analysis of neural networks as networks (strongly connected components)
6. Multiplex networks robustness - Lydia
7. Network analysis of gene ontology as bipartite, as association networks based on terms from real data annotation Konstantinos
8. Inference network from gene expression data and perform the network analysis and gene ontology annotation
9. Integrating genetic and protein–protein interaction networks maps a functional wiring diagram of a cell
10. Analyzing complex networks evolution through Information Theory quantifiers


# Online material

* Network Science [Book](http://networksciencebook.com) by Albert-László Barabási
* Nice network analysis and visualization introduction tutorial [Katherine Ognyanova](http://kateto.net/network-visualization)
* Brief introduction [Introduction to Network Analysis with R](https://www.jessesadler.com/post/network-analysis-with-r/)
* Introduction to [Tidygraph](https://www.data-imaginist.com/2017/introducing-tidygraph/)
* An awesome [list](https://github.com/briatte/awesome-network-analysis) of resources to construct, analyze and visualize network data
* Experiments with [igraph](https://www.r-bloggers.com/experiments-with-igraph/)
* Introduction to Bioconductor [here](https://bioconductor.github.io/BiocWorkshops/introduction-to-bioconductor-annotation-resources.html)
* Epidemic modeling in R, packages, tutorials and workshops : [EpiModel](http://www.epimodel.org/tut.html)
* Simulating network diffusion with R [Tutorial](https://chengjunwang.com/post/en/2014-03-09-simulate-network-diffusion-with-r/)
* Duke Network Analysis Center [Diffusion simulations](https://dnac.ssri.duke.edu/r-labs/2017/04_diffusion_simulations.php)

  